mqtt/mqtt-bridge/tests/integration.rs

```diff
@@ -122,6 +123,89 @@ async fn send_message_upstream_downstream() {
     local_client.shutdown().await;
 }
 
+#[ignore = "This test can be activated once RingBuffer is merged in."]
+#[tokio::test]
+async fn send_message_upstream_with_crash_is_lossless() {
+    let subs = vec![Direction::Out(TopicRule::new(
+        "temp/#".into(),
+        Some("to".into()),
+        Some("upstream".into()),
+    ))];
+
+    let (mut local_server_handle, _, mut upstream_server_handle, _) =
+        common::setup_brokers(AllowAll, AllowAll);
+
+    let mut local_client = TestClientBuilder::new(local_server_handle.address())
+        .with_client_id(ClientId::IdWithExistingSession("local_client".into()))
+        .build();
+
+    let mut upstream_client = TestClientBuilder::new(upstream_server_handle.address())
+        .with_client_id(ClientId::IdWithExistingSession("upstream_client".into()))
+        .build();
+    {
+        let (controller_handle, controller_task) = common::setup_bridge_controller(
+            local_server_handle.address(),
+            upstream_server_handle.tls_address().unwrap(),
+            subs.clone(),
+            "send_message_upstream_with_crash_is_lossless",
+        )
+        .await;
+
+        local_client
+            .subscribe("downstream/filter/#", QoS::AtLeastOnce)
+            .await;
+
+        // wait to receive subscription ack
+        local_client.subscriptions().next().await;
+
+        upstream_client
+            .subscribe("upstream/temp/#", QoS::AtLeastOnce)
+            .await;
+
+        // wait to receive subscription ack
+        upstream_client.subscriptions().next().await;
+
+        // send upstream
+        local_client
+            .publish_qos1("to/temp/1", "from local", false)
+            .await;
+
+        assert_matches!(
+            upstream_client.publications().next().await,
+            Some(ReceivedPublication{payload, .. }) if payload == Bytes::from("from local")
+        );
+
+        // send upstream but disconnect afterwards
+        local_client
+            .publish_qos1("to/temp/1", "from local again", false)
+            .await;
+
+        controller_handle.shutdown();
+        controller_task.await.expect("controller task");
+    }
```

Comment:
Help me understand please what is the benefit of nested scopes here? :)

Reply:
I had done this to force dropping of any resources, from other comment, I see there is a problem with this kind of 'disconnect' and what I really want is to shutdown and then restart the handle like some of the other tests do